### ISSUE TITLE
Alt Text Field in Image Uploads

### ISSUE THREAD

**Issue by erinathop** [ExpressionEngine/ExpressionEngine#1712]
**Is your feature request related to a problem? Please describe.**

The lack of ability to add alternative text to images out-of-the-box with EE is frustrating when trying to make it as easy as possible for clients to add alt text to their images. Simply defaulting to use the file-name isn't very helpful for users interacting with the site using a screen reader. Imagine having to hear a robotic voice read "background_img_99891328_final_final.png" every time an image is described to you.

Furthermore most WCAG browser extensions will read alt text with file-names as "Suspicious alt text". Simply using the description field isn't enough either because of the different values a visual user and a visually impaired user will get from an image description. For example, the use of colours in an image description may be irrelevant to a visually impaired person who's never seen a colour. 

**Describe the solution you'd like**

I'd like an additional field for alt text that is attached to images the way Titles and Descriptions are, as well as a more detailed fall-back if none is provided. Currently alt text defaults to the file-name, however it should look for content in the alt text field first, if the alt text field is empty use the title, and if the title is empty us the image file-name.

**Describe alternatives you've considered**

We've created additional fields in Matrix grids to add alt text to images, as well as the Assets add-on which supports alt text by default.

**Teachability, Documentation, Adoption, Migration Strategy**

Users will use this field just as they would use the "Title" or "Description" field for file uploads. Further reading on the importance and uses of alt text can be found here: https://webaim.org/techniques/alttext/

- [x] I am capable and would like to work on implementation of this feature if it is considered.

**Comment by robsonsobral** [comment]
Hi, @erinathop .

May I disagree just a little?

The `ALT` attribute is intended to use as a description, but not exactly literal.

From [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img):

> The `alt` attribute holds a text description of the image, which isn't mandatory but is **incredibly useful** for accessibility — screen readers read this description out to their users so they know what the image means. Alt text is also displayed on the page if the image can't be loaded for some reason: for example, network errors, content blocking, or linkrot.

From [WCAG](https://www.w3.org/WAI/tutorials/images/informative/):
> The text alternative should convey the meaning or content that is displayed visually, which typically isn’t a literal description of the image.
>
> In some situations a detailed literal description may be needed, but only when the content of the image is all or part of the conveyed information.

I've read some months ago about how visually impaired users like to have a sense of the "mood" of the image, which may includes the color. Sadly, I don't remember where. I gonna update this in case I find it.

So I think the **Description** field serves well to this purpose. (And I don't imagine any other use for it. Do you?)

**Comment by robsonsobral** [comment]
@erinathop , you may be interested in collaborate at [issue 1120](https://github.com/ExpressionEngine/ExpressionEngine/issues/1120)

**Comment by erinathop** [comment]
Hi @robsonsobral, 

You certainly can! What you're saying makes sense, and i don't entirely disagree. There are a couple cases i can think of that would warrant two separate fields.

Going to use instagram/facebook as an example here. Let's say you go to post an image to instagram or facebook, and you add a description to the image - something like "Me and Robin out getting ice cream!". That isn't the most descriptive way to describe the image to a visually impaired person, however the description is still relevant to visual users. The alt text might look like "Two people standing next to an ice cream truck with ice cream cones". 

Another case i can think of is images of text. Think of an image with an inspirational quote in it. It's illegible to a visually impaired person. The description could be telling a story about the quote, adding the title, author and how or why the author came up with the quote in the description. However we still need to add the full quote to the image alt text. 

I hope that makes sense!

**Comment by robsonsobral** [comment]
They are both cases for the [`<figcaption>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figcaption). A caption, unlike the description, can change in context.

Using your example:

```
<figure>
    <img alt="Two people standing next to an ice cream truck with ice cream cones" />
    <figcaption>Me and Robin out getting ice cream!</figcaption>
</figure>
```

Later, on a new publication, the same image can have a completely different meaning:

```
<figure>
    <img alt="Two people standing next to an ice cream truck with ice cream cones" />
    <figcaption>Last vacation the lying douchebag and I have together!</figcaption>
</figure>
```

The file is the same, but the caption is related to the entry which is using it.

Does it make sense?

I think we could improve the label of the **Description** field to **Description to the visually impaired**.

**Comment by erinathop** [comment]
Hmm, yeah, i think you're right. I understand what you mean. 

I think being more descriptive with the labeling of that field would certainly be an improvement. Is there reasoning not to simply label it "Alt text"?

Furthermore, i want to make sure the other part of this isn't being missed, which is the use of file-names as a fallback in place of alternative text. I think we should be checking wether or not a title or alt text/description exists before defaulting to using the file-name. I'm still a little new to expression engine as a whole, so please correct me if i'm wrong, but from what i understand this is not how it's currently being handled.

**Comment by robsonsobral** [comment]
I prefer to not use "ALT Tex" because it's too much of a dev thing. Normal humans don't understand it.

On the other thing, @erinathop , if you don't mind, you can start by helping me [here](https://github.com/ExpressionEngine/ExpressionEngine/pull/1668). I wasn't capable of properly fix it.

**Comment by erinathop** [comment]
Hi @robsonsobral i'll be taking a look into this more over the next couple of weeks, just a bit of internal discussion needed before we proceed!
